actors/ethaccount/src/lib.rs

```diff
@@ -46,67 +55,90 @@ impl EthAccountActor {
 
         Ok(())
     }
-}
 
-impl ActorCode for EthAccountActor {
-    type Methods = Method;
-    actor_dispatch! {
-        Constructor => constructor,
-    }
-}
+    /// Authenticates whether the provided signature is valid for the provided message.
+    /// Should be called with the raw bytes of a signature, NOT a serialized Signature object that includes a SignatureType.
+    /// Errors with USR_ILLEGAL_ARGUMENT if the authentication is invalid.
+    pub fn authenticate_message(
+        rt: &mut impl Runtime,
+        params: AuthenticateMessageParams,
+    ) -> Result<(), ActorError> {
+        rt.validate_immediate_caller_accept_any()?;
+        let msg_hash = rt.hash_blake2b(&params.message);
+
+        let signer_pk = rt
+            .recover_secp_public_key(
+                &msg_hash,
+                params
+                    .signature
+                    .as_slice()
+                    .try_into()
+                    .map_err(|_| actor_error!(illegal_argument; "invalid signature length"))?,
+            )
+            .map_err(|e| {
+                e.downcast_default(
+                    ExitCode::USR_ILLEGAL_ARGUMENT,
+                    "failed to recover signer public key",
+                )
+            })?;
 
-#[cfg(test)]
-mod tests {
-    use fil_actors_runtime::EAM_ACTOR_ID;
-    use fvm_shared::address::Address;
-    use fvm_shared::error::ExitCode;
-    use fvm_shared::MethodNum;
+        // 0x04 to indicate uncompressed point
+        if signer_pk[0] != 0x04 {
+            return Err(actor_error!(assertion_failed; "pubkey should start with 0x04, not {}",
+                signer_pk[0]));
+        }
 
-    use fil_actors_runtime::test_utils::{
-        expect_abort_contains_message, MockRuntime, SYSTEM_ACTOR_CODE_ID,
-    };
-    use fil_actors_runtime::SYSTEM_ACTOR_ADDR;
+        // The subaddress is the last 20 bytes of the keccak hash of the public key
+        let signer_pk_hash = rt.hash(Keccak256, &signer_pk[1..]);
+        if signer_pk_hash.len() < 20 {
+            return Err(
+                actor_error!(assertion_failed; "invalid keccak hash length {}", signer_pk_hash.len()),
+            );
+        }
 
-    use crate::{EthAccountActor, Method};
+        let signer_subaddress_bytes = &signer_pk_hash[signer_pk_hash.len() - 20..];
 
-    const EOA: Address = Address::new_id(1000);
+        let self_address = rt
+            .lookup_delegated_address(
+                rt.message().receiver().id().expect("receiver must be ID address"),
+            )
+            .context_code(
+                ExitCode::USR_ILLEGAL_STATE,
+                "ethaccount should always have delegated address",
+            )?;
 
-    pub fn new_runtime() -> MockRuntime {
-        MockRuntime {
-            receiver: EOA,
-            caller: SYSTEM_ACTOR_ADDR,
-            caller_type: *SYSTEM_ACTOR_CODE_ID,
-            ..Default::default()
+        let self_address_bytes = self_address.payload_bytes();
```

Comment:
I would still double-check the address protocol (just in case we change something).